### DISCUSSION TITLE
How to set the Kyuubi GROUP level on k8s with Spark (JDBC mode)?

### DISCUSSION THREAD

**Discussion by HaoYang670** [apache/kyuubi#4478] [Q&A]
Hi Kyuubi team,

I am currently trying to use Kyuubi on K8s with Spark (by using JDBC through [JayDeBeApi])(https://pypi.org/project/JayDeBeApi/#usage). And I want to set the share level of Kyuubi as GROUP. I've found from the doc that I could set the config `kyuubi.session.engine.share.level=GROUP` to during connecting. However, it fallbacks to USER level because it can't find a group for the user.

My questions are:
1. When and how should I define the relationship between groups and users? I find from the doc that Kyuubi follows the [Hadoop GroupsMapping](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/GroupsMapping.html) to map user to a primary group. Should I setup the `Hadoop Groups Mapping`? If so, how should I do it on K8S?
2. How can I specify the group of a user when connecting a spark cluster? I don't find where I can set it in the JDBC url (https://kyuubi.readthedocs.io/en/v1.6.1-incubating/client/jdbc/kyuubi_jdbc.html#basic-connection-url-format)

**Comment by pan3793** [comment]
Before answer your question, I would suggest you take a look on https://github.com/apache/kyuubi/pull/3897

> ... setup the Hadoop Groups Mapping? If so, how should I do it on K8S?

All things can be done by a `core-site.xml`

**Reply by HaoYang670** [reply]
> Before answer your question, I would suggest you take a look on https://github.com/apache/kyuubi/pull/3897

Thank you @pan3793 . I will go through it.

> kyuubi.session.engine.share.level=GROUP is kind of sessionProperties, just put it in the URL.

Yes, we can set the GROUP level by this. But what if a user belongs to more than one groups? How should I specify the group for that user?

**Reply by pan3793** [reply]
Just FYI, 1.7.0 is available, we are preparing the release notes and will announce it today.

- https://github.com/apache/kyuubi-website/pull/120

**Reply by pan3793** [reply]
> Yes, we can set the GROUP level by this. But what if a user belongs to more than one groups? How should I specify the group for that user?

Sorry, I misread the question the first time. In 1.6.1 and early, it's all Hadoop configuration stuff, since 1.7.0, if you are not familiar w/ [Hadoop GroupsMapping](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/GroupsMapping.html) or think it's too complex, you can easily implement a custom `GroupProvider` to achieve that.

**Reply by HaoYang670** [reply]
Very glad to know this. 

> if you are not familiar w/ [Hadoop GroupsMapping](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/GroupsMapping.html) or think it's too complex, you can easily implement a custom GroupProvider to achieve that.

No I am not. Is there any doc I could follow to implement the JDBC group provider?

**Reply by pan3793** [reply]
> How can I specify the group of a user when connecting a spark cluster? 

A quick idea is implement a `SimpleGroupProvider`, which should be like

```
class SimpleGroupProvider extends GroupProvider {

  override def primaryGroup(user: String, sessionConf: JMap[String, String]): String {
    sessionConf.get("kyuubi.session.group").orElse(user)
  }
}
```

And set ·kyuubi.session.group.provider=xx.xxx.SimpleGroupProvider` in `kyuubi-defaults.conf`, then you can set arbitrary group by setting `kyuubi.session.group=<grp>` in the JDBC url, it's convenient but may has security issue.

If you like, you can send a PR to make it as a built-in option.

**Reply by pan3793** [reply]
> Is there any doc I could follow to implement the JDBC group provider?

Sorry, no particular docs yet. The interface GroupProvider should be simple enough, you just need to implement it and set `kyuubi.session.group.provider=xx.xxx.JdbcGroupProvider`

**Reply by HaoYang670** [reply]
Sorry, it may be my misunderstanding, but I have some questions:
1. Where should I define the `SimpleGroupProvider`?
2. Is `kyuubi.session.group.provider` a session property that could be passed into the JDBC connection URL?

**Reply by HaoYang670** [reply]
Also, I don't find the definition of `JdbcGroupProvider` in your PR 3897. How should I use it?

**Reply by pan3793** [reply]
> 1. Where should I define the SimpleGroupProvider?

Any third jar, or do it in Kyuubi project, anyway, make sure it can be found in Kyuubi Server classpath

> 2. Is kyuubi.session.group.provider a session property that could be passed into the JDBC connection URL?

You should put it in the `kyuubi-defauts.conf`, I updated the previous reply to make it clear.

**Reply by pan3793** [reply]
> Also, I don't find the definition of `JdbcGroupProvider` in your PR #3897. How should I use it?

#3897 does not implement `JdbcGroupProvider`, you should implement it before using

**Comment by HaoYang670** [comment]
Hi, @pan3793 Can a user set his group in JDBC URL (not use the primary group). If so, how could we do this?

**Reply by HaoYang670** [reply]
Hi @pan3793, sorry to bother you again. But could we achieve this in JDBC URL?
